converter/internal/promtailconvert/promtailconvert.go

```diff
@@ -0,0 +1,270 @@
+	if cfg.LimitsConfig != defaultLimitsConfig() {
+		diags.Add(
+			diag.SeverityLevelError,
+			"limits_config is not yet supported in Flow Mode",
+		)
+	}
+
+	// We cannot migrate the tracing config to Flow Mode, since in promtail it relies on
+	// environment variables that can be set or not and depending on what is set, different
+	// features of tracing are configured. We'd need to have conditionals in the
+	// flow config to translate this. See https://www.jaegertracing.io/docs/1.16/client-features/
+	if cfg.Tracing.Enabled {
+		diags.Add(
+			diag.SeverityLevelWarn,
+			"tracing configuration cannot be migrated to Flow Mode automatically - please "+
+				"refer to documentation on how to configure tracing in Flow Mode",
+		)
+	}
+
+	if cfg.TargetConfig.Stdin {
+		diags.Add(
+			diag.SeverityLevelError,
+			"reading targets from stdin is not supported in Flow Mode configuration file",
+		)
+	}
+
+	var writeReceivers = make([]loki.LogsReceiver, len(cfg.ClientConfigs))
+	// Each client config needs to be a separate remote_write,
+	// because they may have different ExternalLabels fields.
+	for i, cc := range cfg.ClientConfigs {
+		writeReceivers[i] = appendLokiWrite(f, &cc, &diags, i)
+	}
+
+	gc := &build.GlobalContext{
+		WriteReceivers:   writeReceivers,
+		TargetSyncPeriod: cfg.TargetConfig.SyncPeriod,
+	}
+
+	for _, sc := range cfg.ScrapeConfig {
+		appendScrapeConfig(f, &sc, &diags, gc)
+	}
+
+	return diags
+}
+
+func defaultPositionsConfig() positions.Config {
+	// We obtain the default by registering the flags
+	cfg := positions.Config{}
+	cfg.RegisterFlags(flag.NewFlagSet("", flag.PanicOnError))
+	return cfg
+}
+
+func defaultLimitsConfig() limit.Config {
+	cfg := limit.Config{}
+	cfg.RegisterFlagsWithPrefix("", flag.NewFlagSet("", flag.PanicOnError))
+	return cfg
+}
+
+func appendScrapeConfig(
+	f *builder.File,
+	cfg *scrapeconfig.Config,
+	diags *diag.Diagnostics,
+	gctx *build.GlobalContext,
+) {
+
+	b := build.NewScrapeConfigBuilder(f, diags, cfg, gctx)
+
+	// Append all the SD components
+	b.AppendKubernetesSDs()
+	//TODO(thampiotr): add support for other SDs
+
+	// Append loki.source.file to process all SD components' targets.
+	// If any relabelling is required, it will be done via a discovery.relabel component.
+	// The files will be watched and the globs in file paths will be expanded using discovery.file component.
+	// The log entries are sent to loki.process if processing is needed, or directly to loki.write components.
+	b.AppendLokiSourceFile()
+
+	// Append all the components that produce logs directly.
+	// If any relabelling is required, it will be done via a loki.relabel component.
+	// The logs are sent to loki.process if processing is needed, or directly to loki.write components.
+	//TODO(thampiotr): add support for other integrations
+	b.AppendCloudFlareConfig()
+	b.AppendJournalConfig()
+}
+
+func appendLokiWrite(f *builder.File, client *client.Config, diags *diag.Diagnostics, index int) loki.LogsReceiver {
```

Comment:
consider now or later breaking these functions up into their own files for each flow component. it's not too bad now but I think it may get less manageable as you continue.

Reply:
makes sense, done!